model/task/task.go

```diff
@@ -376,20 +390,71 @@ func (t *Task) DependenciesMet(depCaches map[string]Task) (bool, error) {
 	if len(depIdsToQueryFor) > 0 {
 		newDeps, err := Find(ByIds(depIdsToQueryFor).WithFields(StatusKey, DependsOnKey))
 		if err != nil {
-			return false, err
+			return nil, errors.WithStack(err)
 		}
 
 		// add queried dependencies to the cache
 		for _, newDep := range newDeps {
 			deps = append(deps, newDep)
-			depCaches[newDep.Id] = newDep
+			depCache[newDep.Id] = newDep
 		}
 	}
 
-	for _, depTask := range deps {
-		if !t.satisfiesDependency(&depTask) {
+	return deps, nil
+}
+
+// DependenciesSatisfiable returns true when a dependency is satisfied
+// or could be satisfied in the future, and false if any of the first
+// order dependencies are blocked, checking the cache if necessary.
+//
+// Use dependencies met (which can also check the cache,) to figure
+// out if a task is runable.
+func (t *Task) DependencySatisfiable(depCache map[string]Task) (bool, error) {
+	if len(t.DependsOn) == 0 || t.OverrideDependencies {
+		return true, nil
+	}
+
+	// do this early, if possible to avoid caching tasks that we
+	// won't need.
+	for _, dep := range t.DependsOn {
+		if dep.Unattainable {
+			return false, nil
+		}
+	}
+
+	deps, err := t.populateDependencyTaskCache(depCache)
+	if err != nil {
+		return false, errors.WithStack(err)
+	}
+
+	for _, dep := range t.DependsOn {
+		depTask, ok := depCache[dep.TaskId]
+		if !ok {
+			grip.Error(message.Fields{
+				"task_id":     t.Id,
+				"dep_task_id": depTask.Id,
+				"cache_size":  len(deps),
+				"operation":   "DependencySatisfiable",
+				"message":     "cache failure",
+			})
+			continue
+		}
+
+		if depTask.Blocked() {
 			return false, nil
 		}
+
+		if !t.satisfiesDependency(&depTask) {
+			if depTask.IsFinished() {
```

Comment:
I would think we'd never get here given that Unattainable was false

Reply:
I think that sometimes data isn't consistent and this is being careful and _extremely_ defensive.